### DISCUSSION TITLE
how to restrict a package to a specific version when using custom manager

### DISCUSSION THREAD

**Discussion by amimas** [renovatebot/renovate#30916] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab: 17.1.2; Renovate: 38

### Please tell us more about your question or problem

Hi,

I have setup a preset containing `customManager` for managing versions in dockerfiles using `ARG` or `ENV`. The config is as shown below. BTW, does it make sense to have this preset be available from renovate directly? I assume everybody needs something like this or maybe it's already available and I missed it 🤷 .

```json5
  customManagers: [
    // Manager that extracts version information from Dockerfile ARG or ENV statements using a custom comment format
    // This allows RenovateBot to manage these for updates.
    {
      customType: "regex",
      fileMatch: [
        // match standard Dockerfile pattern (see: https://docs.renovatebot.com/modules/manager/dockerfile/)
        "(^|/|\\.)Dockerfile$",
        "(^|/)Dockerfile[^/]*$"
      ],
      matchStrings: [
        "#\\srenovate:\\sdatasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?( extractVersion=(?<extractVersion>.*?))?\\s(ENV|ARG) .*?_VERSION=(?<currentValue>.*)\\s"
      ]
    }
  ]
```

This preset is being used in many of my projects and is working well so far. In one project where I applied this has a special requirement. The snippet of dockerfile in that project is as shown below:

```dockerfile
# renovate: datasource=node depName=node
ARG IMAGE_VERSION=22.6.0
# renovate: datasource=npm depName=yarn
ARG YARN_VERSION=1.22.22
# renovate: datasource=docker depName=python
ARG PYTHON3_11_VERSION=3.11.6
```

Notice the last config in the docker file is about managing deps of Python, specifically for version 3.11. The dockerfile uses Amazon Linux 2023 as the base image and python is installed using the command `dnf install python3.11-${PYTHON3_11_VERSION} -y`.

As of right now, Amazon Linux 2023 is not available in repology. That's why the renovate config for the `ARG` is set to look at dockerhub (i.e. `# renovate: datasource=docker depName=python`). The problem is that renovate would occasionally open MR for updating python to version 3.12 because that's available in docker hub. But, this version of python is not yet available in the registry of Amazon Linux 2023. So, we would end up simply closing the renovate MR.

I'm wondering if in this commented string `# renovate: datasource=docker depName=python`, can I specify renovate to stick with the latest of the minor version? Or is there a simpler option?

I hope I explained it clearly. If you need more clarification, please let me know. Thanks.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment, marked as answer]
You can use `allowedVersions`.[^1]

[^1]: https://docs.renovatebot.com/configuration-options/#allowedversions

**Reply by amimas** [reply]
Thanks for the suggestion. How do I target the specific package? For example, I only want the `allowedVersion` to be applied to this `ARG PYTHON3_11_VERSION=3.11.6`.

I'm not sure if the following is valid way of doing this. It seems invalid to me.

```json
{
  "packageRules": [
    {
      "matchPackageNames": ["PYTHON3_11_VERSION"],
      "allowedVersions": "<3.12"
    }
  ]
}
```

**Reply by RahulGautamSingh** [reply]
It is valid.

**Reply by amimas** [reply]
Oh. That's interesting! Didn't think `PYTHON3_11_VERSION` would be considered as a "package name" when using custom manager. I'll give that a shot and see how that goes.

Is this documented somewhere? Wondering what else can I do with the custom manager.

**Reply by RahulGautamSingh** [reply]
My bad the depName will be python as per you custom manger config.

**Reply by amimas** [reply]
I think I get it now. Based on the usage of the custom manager (i.e. `# renovate: datasource=docker depName=python`), I can use `packageRules` to target the appropriate `depName` and then use `allowedVersions` config. Thanks!
